AI/Nullkiller/Analyzers/BuildAnalyzer.cpp

```diff
@@ -30,30 +31,22 @@ void BuildAnalyzer::updateTownDwellings(TownDevelopmentInfo & developmentInfo)
 	for(int level = 0; level < developmentInfo.town->getTown()->creatures.size(); level++)
 	{
 		logAi->trace("Checking dwelling level %d", level);
-		BuildingInfo nextToBuild = BuildingInfo();
-
-		BuildingID buildID = BuildingID(BuildingID::getDwellingFromLevel(level, 0));
-
-		for(; developmentInfo.town->getBuildings().count(buildID); BuildingID::advanceDwelling(buildID))
+		for(BuildingID buildID = BuildingID::getDwellingFromLevel(level, 0); buildID.hasValue(); BuildingID::advanceDwelling(buildID))
 		{
-			if(!developmentInfo.town->hasBuilt(buildID))
+			if(developmentInfo.town->getTown()->buildings.count(buildID) == 0)
 				continue; // no such building in town
 
 			auto info = getBuildingOrPrerequisite(developmentInfo.town, buildID);
 
+			if (!info.exists && !info.canBuild && !info.notEnoughRes)
+				continue;
+
 			if(info.exists)
-			{
 				developmentInfo.addExistingDwelling(info);
+			else
+				developmentInfo.addBuildingToBuild(info);
```

Comment:
this will stop loop at first occurrence of !info.exists, rather than iterating until last one with info.exists is found so value in addBuildingToBuild will be different - what will this change?

Reply:
Should not change anything. This is actually part of the fix.
Compare this with code that was here before it was broken:
https://github.com/vcmi/vcmi/blob/60435a70e2a5615d0f1e49039e78a86d01286f87/AI/Nullkiller/Analyzers/BuildAnalyzer.cpp

It had reversed loop (upgrades 2/1/0), and then - last-most building was picked.
After PR #5174, loop direction has been changed, but it would still pick info from last-most iteration (so AI would attempt to build upgraded dwelling before base).